storage/memory.go

```diff
@@ -0,0 +1,106 @@
+package storage
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/TwinProduction/gatus/core"
+)
+
+var (
+	serviceStatuses = make(map[string]*core.ServiceStatus)
+
+	// serviceResultsMutex is used to prevent concurrent map access
+	serviceResultsMutex sync.RWMutex
+)
+
+// InMemoryStore implements an in-memory store
+type InMemoryStore struct{}
+
+// NewInMemoryStore returns an in-memory store. Note that the store acts as a singleton, so although new-ing
+// up in-memory stores will give you a unique reference to a struct each time, all structs returned
+// by this function will act on the same in-memory store.
+func NewInMemoryStore() InMemoryStore {
+	return InMemoryStore{}
+}
+
+// GetAll returns all the observed results for all services from the in memory store
+func (ims *InMemoryStore) GetAll() map[string]*core.ServiceStatus {
+	results := make(map[string]*core.ServiceStatus)
+	serviceResultsMutex.RLock()
+	for key, svcStatus := range serviceStatuses {
+		copiedResults := copyResults(svcStatus.Results)
+		results[key] = &core.ServiceStatus{
+			Name:    svcStatus.Name,
+			Group:   svcStatus.Group,
+			Results: copiedResults,
+		}
```

Comment:
Could you just create a new map here without copying the results?
Since the only use for GetAll is to Marshal it to json, it shouldn't be necessary to copy everything.
```suggestion
		results[key] = svcStatus
```

Alternatively, you could just remove the `GetAll` function and add a function for marshaling the serviceStatuses, that way, you wouldn't need to copy anything and you could just `json.Marshal` it while the mutex is locked.

Up to you.

Reply:
The benefit of returning a copy means that the internals of the store are entirely decoupled from the consumer.

Whilst it's true that "the only use for GetAll is to Marshal it to json" _for now_, that defeats the point of having this abstraction, as the decoupling its providing is broken because of the implicit assumption. In general, it's a dangerous pattern to assume how the consumer uses returned value whenever it's a reference. 

IMO, it's much better to be defensive (e.g, via this copying) as then there's no concern for how the internal map is consumed after it's returned to the caller.

I'm also not a fan of making this return the marshalled JSON, as it's then no longer a simple Memory Store, but has an additional responsibility (marshalling the data) which I don't believe it owns.

Reply:
Considering there's no external consumers and it's only used internally, making this kind of assumption wouldn't be _that_ much of a stretch, but fair enough. My only worry was unnecessary memory allocations coming from constantly calling `GetAll`, which duplicates the entire map.

Reply:
Hopefully the `pprof` output puts some of those concerns at ease? Admittedly its only 5m of profiling, but the overhead is indistinguishable so I don't think this is a concern... 

The structs themselves are also incredibly simple so I think avoiding the allocation is a premature optimisation at the cost of safety.

Reply:
During the 5 minutes, did you have the dashboard up? Only by having the dashboard up will the `GetAll` be called.

There's also a ~10 second period where the JSON for the service status is cached to prevent constantly re-marshaling + compressing the output with gzip everything when multiple clients are calling `/api/v1/statuses` concurrently.

If you could just add the following in the main while running pprof for 5 minutes, you'd be able to get a much more accurate idea of what allocations will look like:

```go
	go func() {
		for {
			store.GetAll()
			time.Sleep(10*time.Second)
		}
	}()
```

Alternatively, you could go on the dashboard and set the refresh interval to 10s:
![image](https://user-images.githubusercontent.com/15699766/103426356-d3708680-4b86-11eb-835d-ac82710655f5.png)

Reply:
Here's the output after running it for 5 minutes with the repeated polling of `store.Getall()` every 10s as suggested:

Allocs:

```
# runtime.MemStats
# Alloc = 3101272
# TotalAlloc = 81223600
# Sys = 73482496
# Lookups = 0
# Mallocs = 225421
# Frees = 202183
# HeapAlloc = 3101272
# HeapSys = 66519040
# HeapIdle = 61792256
# HeapInuse = 4726784
# HeapReleased = 61497344
# HeapObjects = 23238
# Stack = 589824 / 589824
# MSpan = 96016 / 114688
# MCache = 6944 / 16384
# BuckHashSys = 1452488
# GCSys = 3574024
# OtherSys = 1216048
# NextGC = 4194304
# LastGC = 1609451401253367000
# PauseNs = [38287 392749 33183 91522 66646 119242 92871 76613 145293 69021 25624 196362 32264 57675 33916 93387 141996 84793 69983 46402 107785 122480 59423 274082 73875 65611 45099 150205 43665 53501 178500 3613864 568936 30873 36352 266245 65896 70915 65341 82028 28818 54278 96434 63834 46213 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
# PauseEnd = [1609451102330061000 1609451109795588000 1609451114619471000 1609451121047953000 1609451127730080000 1609451134584871000 1609451141893116000 1609451149701862000 1609451156283084000 1609451162955165000 1609451169735606000 1609451177234012000 1609451183920375000 1609451191269862000 1609451198181284000 1609451204791207000 1609451212112645000 1609451219150862000 1609451225833999000 1609451233424987000 1609451240160535000 1609451246841568000 1609451254234123000 1609451261007467000 1609451267590011000 1609451274271590000 1609451281670695000 1609451288353576000 1609451294935018000 1609451301711944000 1609451308694865000 1609451315613165000 1609451322569516000 1609451328959417000 1609451334601577000 1609451340107482000 1609451346234744000 1609451353259476000 1609451359424001000 1609451366618580000 1609451373780183000 1609451380379298000 1609451387369216000 1609451394143047000 1609451401253367000 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
# NumGC = 45
# NumForcedGC = 0
# GCCPUFraction = 0.00011319083956920442
# DebugGC = false
```

Heaps:

```
# runtime.MemStats
# Alloc = 2386224
# TotalAlloc = 83898328
# Sys = 73482496
# Lookups = 0
# Mallocs = 233247
# Frees = 211990
# HeapAlloc = 2386224
# HeapSys = 66519040
# HeapIdle = 62627840
# HeapInuse = 3891200
# HeapReleased = 61833216
# HeapObjects = 21257
# Stack = 589824 / 589824
# MSpan = 98872 / 114688
# MCache = 6944 / 16384
# BuckHashSys = 1452488
# GCSys = 3574024
# OtherSys = 1216048
# NextGC = 4194304
# LastGC = 1609451413968758000
# PauseNs = [38287 392749 33183 91522 66646 119242 92871 76613 145293 69021 25624 196362 32264 57675 33916 93387 141996 84793 69983 46402 107785 122480 59423 274082 73875 65611 45099 150205 43665 53501 178500 3613864 568936 30873 36352 266245 65896 70915 65341 82028 28818 54278 96434 63834 46213 78697 2102388 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
# PauseEnd = [1609451102330061000 1609451109795588000 1609451114619471000 1609451121047953000 1609451127730080000 1609451134584871000 1609451141893116000 1609451149701862000 1609451156283084000 1609451162955165000 1609451169735606000 1609451177234012000 1609451183920375000 1609451191269862000 1609451198181284000 1609451204791207000 1609451212112645000 1609451219150862000 1609451225833999000 1609451233424987000 1609451240160535000 1609451246841568000 1609451254234123000 1609451261007467000 1609451267590011000 1609451274271590000 1609451281670695000 1609451288353576000 1609451294935018000 1609451301711944000 1609451308694865000 1609451315613165000 1609451322569516000 1609451328959417000 1609451334601577000 1609451340107482000 1609451346234744000 1609451353259476000 1609451359424001000 1609451366618580000 1609451373780183000 1609451380379298000 1609451387369216000 1609451394143047000 1609451401253367000 1609451407346064000 1609451413968758000 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
# NumGC = 47
# NumForcedGC = 0
# GCCPUFraction = 0.0001332981632864582
# DebugGC = false
```

Reply:
I assume that instead of `Allocs` and `Heaps`, you meant `Before` and `After`? 🤣 

I'm honestly at loss for words as to why there's more allocations before vs after, but oh well 🤔

Reply:
Nope, this is the output of the `net/http/pprof` package which breaks down `alloc`'s separately from `heap`'s... 

As described in [this article](https://www.freecodecamp.org/news/how-i-investigated-memory-leaks-in-go-using-pprof-on-a-large-codebase-4bec4325e192/) as:

> heap — a sampling of memory allocations of live objects
> allocs — a sampling of all past memory allocations

Wasn't sure which was more useful, so I included both.